pkg/controller/pinnedimageset/pinned_image_set.go

```diff
@@ -0,0 +1,399 @@
+// It enforces that the syncHandler is never invoked concurrently with the same key.
+func (ctrl *Controller) worker() {
+	for ctrl.processNextWorkItem() {
+	}
+}
+
+func (ctrl *Controller) processNextWorkItem() bool {
+	key, quit := ctrl.queue.Get()
+	if quit {
+		return false
+	}
+	defer ctrl.queue.Done(key)
+
+	err := ctrl.syncHandler(key.(string))
+	ctrl.handleErr(err, key)
+
+	return true
+}
+
+func (ctrl *Controller) handleErr(err error, key interface{}) {
+	if err == nil {
+		ctrl.queue.Forget(key)
+		return
+	}
+
+	if ctrl.queue.NumRequeues(key) < maxRetries {
+		klog.V(2).Infof("Error syncing machineconfigpool %v: %v", key, err)
+		ctrl.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	klog.V(2).Infof("Dropping machineconfigpool %q out of the queue: %v", key, err)
+	ctrl.queue.Forget(key)
+	ctrl.queue.AddAfter(key, 1*time.Minute)
+}
+
+// syncMachineConfigPool will sync the machineconfig pool with the given key.
+// This function is not meant to be invoked concurrently with the same key.
+func (ctrl *Controller) syncMachineConfigPool(key string) error {
+	startTime := time.Now()
+	klog.V(4).Infof("Started syncing machineconfigpool %q (%v)", key, startTime)
+	defer func() {
+		klog.V(4).Infof("Finished syncing machineconfigpool %q (%v)", key, time.Since(startTime))
+	}()
+
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	mcp, err := ctrl.mcpLister.Get(name)
+	if errors.IsNotFound(err) {
+		klog.V(2).Infof("MachineConfigPool %v has been deleted", key)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	pool := mcp.DeepCopy()
+	everything := metav1.LabelSelector{}
+	if reflect.DeepEqual(pool.Spec.MachineConfigSelector, &everything) {
+		return nil
+	}
+
+	selector, err := metav1.LabelSelectorAsSelector(pool.Spec.MachineConfigSelector)
+	if err != nil {
+		return err
+	}
+
+	imageSets, err := ctrl.imageSetLister.List(selector)
+	if err != nil {
+		return err
+	}
+	sort.SliceStable(imageSets, func(i, j int) bool { return imageSets[i].Name < imageSets[j].Name })
+
+	if err := ctrl.syncPinnedImageSets(pool, imageSets); err != nil {
+		klog.Errorf("Error syncing pinned image sets: %v", err)
+		return ctrl.syncFailingStatus(pool, err)
+	}
+
+	return ctrl.syncAvailableStatus(pool)
+}
+
+func (ctrl *Controller) syncAvailableStatus(pool *mcfgv1.MachineConfigPool) error {
+	// TODO: update with the correct condition MachineConfigPoolPinnedImageSetsDegraded
+	if apihelpers.IsMachineConfigPoolConditionFalse(pool.Status.Conditions, mcfgv1.MachineConfigPoolRenderDegraded) {
+		return nil
+	}
+	sdegraded := apihelpers.NewMachineConfigPoolCondition(mcfgv1.MachineConfigPoolRenderDegraded, corev1.ConditionFalse, "", "")
+	apihelpers.SetMachineConfigPoolCondition(&pool.Status, *sdegraded)
+	if _, err := ctrl.client.MachineconfigurationV1().MachineConfigPools().UpdateStatus(context.TODO(), pool, metav1.UpdateOptions{}); err != nil {
```

Comment:
So eventually all of these should be updating the `MachineConfigPoolPinnedImageSetsDegraded` and you're just using renderdegraded for testing?

Reply:
yeah requires an api bump will do that